### ISSUE TITLE
Typing regression in OCaml 4.11 (and trunk)

### ISSUE THREAD

**Issue by nojb** [ocaml/ocaml#9970]
```
type 'a t = ..
type 'a t += X of {x: 'a}
let x: 'a. 'a -> 'a t = fun x -> X {x}
```
gives
```
File "/home/nojebar/foo.ml", line 3, characters 36-37:
3 | let x: 'a. 'a -> 'a t = fun x -> X {x}
                                        ^
Error: The record field x belongs to the type 'b X
       but is mixed here with fields of type 'a X
       The type constructor X would escape its scope
```
The issue is present in 4.11 and trunk (works fine in 4.10).

**Comment by gasche** [comment]
Note: the trunk toplevel behaves in a very weird way, in that if you give the same declaration a *second* time, this time it works.

```
type 'a t = ..
type 'a t += X of { x : 'a; }
# let x: 'a. 'a -> 'a t = fun x -> X {x};;
Error: The record field x belongs to the type 'b X
       but is mixed here with fields of type 'a X
       The type constructor X would escape its scope
# let x: 'a. 'a -> 'a t = fun x -> X {x};;
val x : 'a -> 'a t = <fun>
```

This clearly shows that this is a bug (and not some odd-but-explainable change in unspecified interactions between inference variables and inner-record types). 

Finally, there is no need for a polymorphic annotation, the following is enough:

```
# let x = fun (x : 'a) -> X {x};;
Error: The record field x belongs to the type 'a X
       but is mixed here with fields of type 'b X
       The type constructor X would escape its scope
```
(so this is really about the interaction of inference variables and inner-record types)

**Comment by nojb** [comment]
Friendly ping. Bisect points to #1132 cc @lpw25

**Comment by lpw25** [comment]
Is that the bisect on the original example or on @gasche's version? I take the existence of the second one to mean that the relationship with #1132 is coincidental.

**Comment by nojb** [comment]
> Is that the bisect on the original example or on @gasche's version? 

I used the original example.

**Comment by nojb** [comment]
I am rerunning the bisect on @gasche's version.

**Comment by yallop** [comment]
The problem @gasche demonstrates is present in 4.10.0.

**Comment by yallop** [comment]
> The problem @gasche demonstrates is present in 4.10.0

It was apparently introduced around 4.08.0, since 4.08.0 has the problem but 4.07.0 doesn't.

**Comment by garrigue** [comment]
Note that 4.08.0 corresponds to the introduction of scopes inside identifiers, so this may be the cause.
Pong @trefis

**Comment by nojb** [comment]
Friendly ping. This bug is blocking us from migrating to 4.11.

If I have time, I will try to bisect in the coming week to try to find out the precise commit or PR in 4.08 that introduced it.

**Comment by gasche** [comment]
I tried to hunt this problem as an exercise in understanding inference levels and stuff (cc @garrigue @trefis). There is no interaction with explicit type variables, but there is an interaction with scopes indeed. I think the declaration of extension constructors is currently broken, but I don't know how to fix it.

Repro case:
```
# type _ t = ..;;
# type 'a t += X of {x : 'a};;
# X {x = 1};;
Error: The record field x belongs to the type 'a X
       but is mixed here with fields of type 'b X
       The type constructor X would escape its scope
# X {x = 1};;
- : int t = X {x = <poly>}
```

I understand the following:
- `Typedecl.transl_extension_constructor` creates a new scope for X.
  (In my debug runs, X has scope 3.)
- Then the following toplevel phrase runs at a level below the scope of X. It therefore builds type that are ill-scoped
  (In my debug runs, a Tconstr(X, ...) at level 2).
- When a phrase fails to type-check in the toplevel, the current level is not reset to its previous level.
  The following phrase thus starts with a higher level, higher than the scope of X, so there is no scope escape anymore.
  (In my debug runs, at level 5)

**Comment by gasche** [comment]
Ah, I think I see what to do: the scope should be declared outside the `begin_def..end_def` block. I will try that and send a PR if it passes enough tests.

**Comment by gasche** [comment]
A fix has been merged in trunk in 4.12. We could consider backporting to older releases (the issue we are fixing has been introduced in 4.08, I believe).

**Comment by nojb** [comment]
> A fix has been merged in trunk in 4.12. We could consider backporting to older releases (the issue we are fixing has been introduced in 4.08, I believe).

I think this would be a good idea.
